File: android/src/main/java/PushNotificationsPlugin.kt
// FCM push registration + notification events.
//
// Wire contract mirrors the iOS implementation exactly (payloads match
// src/models.rs / guest-js):
//
//   - `registerForPush` resolves the FCM registration token. The CONSUMER
//     app must be Firebase-configured (google-services.json + the
//     com.google.gms.google-services Gradle plugin) — without it the token
//     fetch rejects at runtime, never at build time.
//   - Foreground deliveries (both data-only and notification messages while
//     the app is in front) emit `notificationReceived`.
//   - Taps emit `notificationTapped`: a background notification-message tap
//     relaunches/resumes the activity with the push's data keys as intent
//     extras, read from the cold-start intent in `load` and from
//     `onNewIntent` after.
//   - Events fired before a JS listener exists are queued and flushed by
//     `startNotificationEvents` (the guest bindings call it right after the
//     first listener registers) — same drain-then-arm contract as iOS.
//
// All mutable state is confined to the main thread (`handler.post`), so the
// plugin needs no further synchronization.

package app.tauri.pushnotifications

import android.Manifest
import android.app.Activity
import android.app.AlarmManager
import android.app.NotificationManager
import android.app.PendingIntent
import android.content.Context
import android.content.Intent
import android.os.Build
import android.os.Handler
import android.os.Looper
import android.webkit.WebView
import app.tauri.PermissionState
import app.tauri.annotation.Command
import app.tauri.annotation.InvokeArg
import app.tauri.annotation.Permission
import app.tauri.annotation.PermissionCallback
import app.tauri.annotation.TauriPlugin
import app.tauri.plugin.Invoke
import app.tauri.plugin.JSObject
import app.tauri.plugin.Plugin
import com.google.firebase.messaging.FirebaseMessaging
import com.google.firebase.messaging.RemoteMessage

private const val PERMISSION_ALIAS = "postNotification"

// Local-notification command payloads (mirror src/models.rs / guest-js).

@InvokeArg
class ScheduleLocalArgs {
    var id: Int = 0
    lateinit var title: String
    var body: String? = null
    var atMs: Double = 0.0
    var data: Map<String, String>? = null
}

@InvokeArg
class CancelLocalArgs {
    var ids: List<Int> = emptyList()
}

@TauriPlugin(
    permissions = [
        Permission(strings = [Manifest.permission.POST_NOTIFICATIONS], alias = PERMISSION_ALIAS)
    ]
)
class PushNotificationsPlugin(private val activity: Activity) : Plugin(activity) {

    companion object {
        // The FCM service (a system-instantiated component) delivers through
        // here; a Tauri app registers exactly one plugin instance. Messages
        // arriving before the plugin loads are queued statically.
        @Volatile internal var instance: PushNotificationsPlugin? = null
        internal val earlyMessages = ArrayDeque<RemoteMessage>()
    }

    private val handler = Handler(Looper.getMainLooper())
    private var armed = false
    private val queuedReceives = mutableListOf<JSObject>()
    private val queuedTaps = mutableListOf<JSObject>()

    override fun load(webView: WebView) {
        super.load(webView)
        instance = this
        handler.post {
            // A cold-start tap launched the activity with the push's data
            // keys as intent extras — queue it for the first listener.
            activity.intent?.let { readTapIntent(it) }
            while (earlyMessages.isNotEmpty()) {
                deliverForegroundMessage(earlyMessages.removeFirst())
            }
        }
    }

    override fun onNewIntent(intent: Intent) {
        handler.post { readTapIntent(intent) }
    }

    // MARK: Commands

    @Command
    fun isPermissionGranted(invoke: Invoke) {
        invoke.resolve(JSObject().put("granted", notificationsAllowed()))
    }

    @Command
    fun requestPermission(invoke: Invoke) {
        if (notificationsAllowed()) {
            invoke.resolve(JSObject().put("granted", true))
            return
        }
        requestPermissionForAlias(PERMISSION_ALIAS, invoke, "permissionCallback")
    }

    @PermissionCallback
    fun permissionCallback(invoke: Invoke) {
        invoke.resolve(JSObject().put("granted", notificationsAllowed()))
    }

    @Command
    fun registerForPush(invoke: Invoke) {
        FirebaseMessaging.getInstance().token
            .addOnSuccessListener { token -> invoke.resolve(JSObject().put("token", token)) }
            .addOnFailureListener { e ->
                invoke.reject("push registration failed: ${e.message ?: "unknown"}")
            }
    }

    /** Arm live delivery and flush everything queued before a JS listener
     *  existed (trigger() drops events with no registered channel). */
    @Command
    fun startNotificationEvents(invoke: Invoke) {
        handler.post {
            armed = true
            queuedReceives.forEach { trigger("notificationReceived", it) }
            queuedReceives.clear()
            queuedTaps.forEach { trigger("notificationTapped", it) }
            queuedTaps.clear()
            invoke.resolve()
        }
    }

    // MARK: Local scheduled notifications (reminders)

    /** Schedule an OS-local notification via AlarmManager. Same-id scheduling
     *  replaces the pending alarm (FLAG_UPDATE_CURRENT on a same-requestCode
     *  PendingIntent) — the resync primitive. Inexact-while-idle: minute-ish
     *  precision without the SCHEDULE_EXACT_ALARM policy surface. Alarms do
     *  not survive reboot; callers resync on next launch. */
    @Command
    fun scheduleLocal(invoke: Invoke) {
        val id = invoke.parseArgs(ScheduleLocalArgs::class.java)
        val alarm = activity.getSystemService(Context.ALARM_SERVICE) as AlarmManager
        alarm.setAndAllowWhileIdle(
            AlarmManager.RTC_WAKEUP,
            id.atMs.toLong(),
            publisherIntent(id.id) { intent ->
                intent.putExtra(EXTRA_TITLE, id.title)
                id.body?.let { intent.putExtra(EXTRA_BODY, it) }
                id.data?.let { intent.putExtra(EXTRA_DATA, HashMap(it)) }
            },
        )
        ledgerUpdate(activity) { it.add(id.id.toString()) }
        invoke.resolve()
    }

    @Command
    fun cancelLocal(invoke: Invoke) {
        val args = invoke.parseArgs(CancelLocalArgs::class.java)
        val alarm = activity.getSystemService(Context.ALARM_SERVICE) as AlarmManager
        val manager =
            activity.getSystemService(Context.NOTIFICATION_SERVICE) as NotificationManager
        for (id in args.ids) {
            alarm.cancel(publisherIntent(id) {})
            manager.cancel(id)
        }
        ledgerUpdate(activity) { set -> args.ids.forEach { set.remove(it.toString()) } }
        invoke.resolve()
    }

    /** Scheduled-but-not-fired ids from the SharedPreferences ledger (Android
     *  has no AlarmManager query API). Reboot drops the alarms but not the
     *  ledger — callers treating this as advisory should resync anyway. */
    @Command
    fun getPendingLocal(invoke: Invoke) {
        val prefs = activity.getSharedPreferences(LEDGER_PREFS, Context.MODE_PRIVATE)
        val ids = prefs.getStringSet(LEDGER_KEY, emptySet())!!.mapNotNull { it.toIntOrNull() }
        val payload = JSObject()
        payload.put("ids", org.json.JSONArray(ids))
        invoke.resolve(payload)
    }

    private fun publisherIntent(id: Int, fill: (Intent) -> Unit): PendingIntent {
        val intent = Intent(activity, LocalNotificationPublisher::class.java)
        intent.putExtra(EXTRA_ID, id)
        fill(intent)
        return PendingIntent.getBroadcast(
            activity,
            id,
            intent,
            PendingIntent.FLAG_UPDATE_CURRENT or PendingIntent.FLAG_IMMUTABLE,
        )
    }

    // MARK: Delivery (from PushNotificationsService)

    internal fun deliverForegroundMessage(message: RemoteMessage) {
        val payload = JSObject()
        message.notification?.title?.let { payload.put("title", it) }
        message.notification?.body?.let { payload.put("body", it) }
        val data = JSObject()
        for ((key, value) in message.data) data.put(key, value)
        payload.put("data", data)
        handler.post { emitOrQueue("notificationReceived", payload, queuedReceives) }
    }

    // MARK: Internals (main thread only)

    private fun notificationsAllowed(): Boolean {
        // Below 13 notifications need no runtime permission.
        if (Build.VERSION.SDK_INT < 33) return true
        return getPermissionState(PERMISSION_ALIAS) == PermissionState.GRANTED
    }

    private fun emitOrQueue(event: String, payload: JSObject, queue: MutableList<JSObject>) {
        if (armed) trigger(event, payload) else queue.add(payload)
    }

    private fun readTapIntent(intent: Intent) {
        val extras = intent.extras ?: return
        // Only intents FCM stamped or our own local-notification marker — a
        // plain launch/deep-link intent is not a notification tap.
        if (extras.getString("google.message_id") == null &&
            extras.getString(LOCAL_TAP_MARKER) == null
        ) {
            return
        }
        val data = JSObject()
        for (key in extras.keySet()) {
            if (key.startsWith("google.") || key.startsWith("gcm.") ||
                key.startsWith("app.tauri.pushnotifications") ||
                key == "from" || key == "collapse_key"
            ) {
                continue
            }
            extras.getString(key)?.let { data.put(key, it) }
        }
        val payload = JSObject().put("data", data)
        emitOrQueue("notificationTapped", payload, queuedTaps)
    }
}

File: ios/Sources/PushNotificationsPlugin.swift
//
//  PushNotificationsPlugin.swift
//  tauri-plugin-push-notifications
//
//  APNs push registration + notification events.
//
//  The app needs the Push Notifications capability (`aps-environment`
//  entitlement); registration rejects without it. Foreground pushes emit
//  `notificationReceived` (and are NOT presented by the system — the app
//  owns its in-app surface); taps emit `notificationTapped`, including the
//  cold-start tap that launched the app.
//
//  Delivery contract: `trigger()` reaches registered JS channels only, so
//  every event is queued until the guest bindings call
//  `startNotificationEvents` (right after the first listener registers),
//  then queued events replay in order. All shared state is confined to the
//  main queue — command handlers hop onto it and never block it.
//

import Foundation
import Tauri
import UIKit
import UserNotifications
import WebKit

// MARK: - Payloads (mirror src/models.rs / guest-js exactly)

struct PermissionPayload: Encodable {
    let granted: Bool
}

struct RegistrationPayload: Encodable {
    let token: String
}

struct NotificationPayload: Encodable {
    let title: String?
    let body: String?
    /// The push's custom key/value bag (everything outside `aps`), with
    /// non-string scalars stringified — deep-link routing data.
    let data: [String: String]
}

struct ScheduleLocalArgs: Decodable {
    /// Stable i32 — re-scheduling the same id replaces the pending request.
    let id: Int
    let title: String
    let body: String?
    /// Epoch milliseconds to fire at; past instants fire ~immediately.
    let atMs: Double
    /// Delivered back through `notificationTapped` as the `data` bag.
    let data: [String: String]?
}

struct CancelLocalArgs: Decodable {
    let ids: [Int]
}

struct PendingLocalPayload: Encodable {
    let ids: [Int]
}

@available(iOS 15.0, *)
class PushNotificationsPlugin: Plugin, UNUserNotificationCenterDelegate {
    /// The instance the swizzled AppDelegate hooks report back to. A Tauri
    /// app registers exactly one instance of a plugin.
    private static weak var shared: PushNotificationsPlugin?

    /// `registerForPush` invokes awaiting the APNs token callback.
    private var pendingRegistrations: [Invoke] = []

    /// Events that fired before JS armed delivery (`startNotificationEvents`)
    /// — most importantly the cold-start tap that launched the app.
    private var armed = false
    private var queuedReceives: [NotificationPayload] = []
    private var queuedTaps: [NotificationPayload] = []

    public override func load(webview: WKWebView) {
        PushNotificationsPlugin.shared = self
        // The notification-center delegate must exist before the first
        // delivery — including the cold-start tap, which UNUserNotification-
        // Center replays shortly after launch to whoever the delegate is.
        // Plugin load runs during application startup, early enough.
        UNUserNotificationCenter.current().delegate = self
        Self.installAppDelegateHooks()
    }

    // MARK: Commands

    @objc public func isPermissionGranted(_ invoke: Invoke) {
        UNUserNotificationCenter.current().getNotificationSettings { settings in
            invoke.resolve(PermissionPayload(granted: settings.authorizationStatus.grantsDelivery))
        }
    }

    @objc public func requestPermission(_ invoke: Invoke) {
        UNUserNotificationCenter.current().requestAuthorization(options: [
            .alert, .badge, .sound,
        ]) { granted, _ in
            // A denied prompt is an answer, not an error.
            invoke.resolve(PermissionPayload(granted: granted))
        }
    }

    /// Ask APNs for the device token. Resolution happens in the swizzled
    /// AppDelegate callbacks below; parking the Invoke (instead of blocking)
    /// keeps the ipc queue free while APNs round-trips.
    @objc public func registerForPush(_ invoke: Invoke) {
        DispatchQueue.main.async { [weak self] in
            guard let self = self else {
                invoke.reject("plugin is gone")
                return
            }
            UNUserNotificationCenter.current().getNotificationSettings { settings in
                guard settings.authorizationStatus.grantsDelivery else {
                    invoke.reject("notification permission not granted")
                    return
                }
                DispatchQueue.main.async {
                    self.pendingRegistrations.append(invoke)
                    UIApplication.shared.registerForRemoteNotifications()
                }
            }
        }
    }

    /// Called by the guest bindings right after a JS listener registers:
    /// replay everything queued while no channel existed, then deliver live.
    @objc public func startNotificationEvents(_ invoke: Invoke) {
        DispatchQueue.main.async { [weak self] in
            guard let self = self else {
                invoke.resolve()
                return
            }
            self.armed = true
            for payload in self.queuedReceives {
                try? self.trigger("notificationReceived", data: payload)
            }
            self.queuedReceives.removeAll()
            for payload in self.queuedTaps {
                try? self.trigger("notificationTapped", data: payload)
            }
            self.queuedTaps.removeAll()
            invoke.resolve()
        }
    }

    // MARK: Local scheduled notifications (reminders)

    /// Schedule an OS-local notification. Same-id scheduling REPLACES the
    /// pending request (UNUserNotificationCenter semantics) — callers resync
    /// by re-scheduling with deterministic ids. Taps flow through the same
    /// `notificationTapped` event as push taps (`didReceive` reads userInfo);
    /// foreground firings keep the system banner (`willPresent` only mutes
    /// PUSH presentations).
    @objc public func scheduleLocal(_ invoke: Invoke) throws {
        let args = try invoke.parseArgs(ScheduleLocalArgs.self)
        let content = UNMutableNotificationContent()
        content.title = args.title
        if let body = args.body { content.body = body }
        if let data = args.data { content.userInfo = data }
        content.sound = .default
        let seconds = max(1, (args.atMs / 1000) - Date().timeIntervalSince1970)
        let trigger = UNTimeIntervalNotificationTrigger(timeInterval: seconds, repeats: false)
        let request = UNNotificationRequest(
            identifier: String(args.id), content: content, trigger: trigger)
        UNUserNotificationCenter.current().add(request) { error in
            if let error = error {
                invoke.reject("scheduling failed: \(error.localizedDescription)")
            } else {
                invoke.resolve()
            }
        }
    }

    @objc public func cancelLocal(_ invoke: Invoke) throws {
        let args = try invoke.parseArgs(CancelLocalArgs.self)
        let ids = args.ids.map(String.init)
        let center = UNUserNotificationCenter.current()
        center.removePendingNotificationRequests(withIdentifiers: ids)
        center.removeDeliveredNotifications(withIdentifiers: ids)
        invoke.resolve()
    }

    /// Ids of still-pending local requests. Only numeric identifiers are
    /// ours (push notifications never enter the pending store).
    @objc public func getPendingLocal(_ invoke: Invoke) {
        UNUserNotificationCenter.current().getPendingNotificationRequests { requests in
            let ids = requests.compactMap { Int($0.identifier) }
            invoke.resolve(PendingLocalPayload(ids: ids))
        }
    }

    // MARK: APNs registration callbacks (via the AppDelegate hooks)

    fileprivate func apnsTokenReceived(_ deviceToken: Data) {
        let token = deviceToken.map { String(format: "%02x", $0) }.joined()
        let waiting = pendingRegistrations
        pendingRegistrations.removeAll()
        for invoke in waiting {
            invoke.resolve(RegistrationPayload(token: token))
        }
    }

    fileprivate func apnsRegistrationFailed(_ error: Error) {
        let waiting = pendingRegistrations
        pendingRegistrations.removeAll()
        for invoke in waiting {
            invoke.reject("push registration failed: \(error.localizedDescription)")
        }
    }

    // MARK: UNUserNotificationCenterDelegate

    /// A push arrived while the app is FOREGROUND. Hand it to JS (the app
    /// owns its in-app surface — a toast/banner of its own) and present
    /// nothing system-side, so the user never sees a double notification.
    /// Non-push notifications (local — not produced by this plugin, but a
    /// host app may schedule its own) keep the system presentation.
    public func userNotificationCenter(
        _ center: UNUserNotificationCenter,
        willPresent notification: UNNotification,
        withCompletionHandler completionHandler: @escaping (UNNotificationPresentationOptions) ->
            Void
    ) {
        guard notification.request.trigger is UNPushNotificationTrigger else {
            completionHandler([.banner, .list, .sound])
            return
        }
        let payload = Self.payload(from: notification.request.content)
        DispatchQueue.main.async { [weak self] in
            self?.emitOrQueue("notificationReceived", payload, queue: \.queuedReceives)
        }
        completionHandler([])
    }

    /// The user TAPPED a notification (background, tray, or the cold-start
    /// tap that launched the app — the center replays that one to the
    /// delegate set at launch, which `load` guarantees we are).
    public func userNotificationCenter(
        _ center: UNUserNotificationCenter,
        didReceive response: UNNotificationResponse,
        withCompletionHandler completionHandler: @escaping () -> Void
    ) {
        if response.actionIdentifier == UNNotificationDefaultActionIdentifier {
            let payload = Self.payload(from: response.notification.request.content)
            DispatchQueue.main.async { [weak self] in
                self?.emitOrQueue("notificationTapped", payload, queue: \.queuedTaps)
            }
        }
        completionHandler()
    }

    /// Main-queue only. Deliver live once armed; queue before that.
    private func emitOrQueue(
        _ event: String,
        _ payload: NotificationPayload,
        queue: ReferenceWritableKeyPath<PushNotificationsPlugin, [NotificationPayload]>
    ) {
        if armed {
            try? trigger(event, data: payload)
        } else {
            self[keyPath: queue].append(payload)
        }
    }

    private static func payload(from content: UNNotificationContent) -> NotificationPayload {
        var data: [String: String] = [:]
        for (key, value) in content.userInfo {
            guard let key = key as? String, key != "aps" else { continue }
            if let string = value as? String {
                data[key] = string
            } else if let number = value as? NSNumber {
                data[key] = number.stringValue
            }
        }
        return NotificationPayload(
            title: content.title.isEmpty ? nil : content.title,
            body: content.body.isEmpty ? nil : content.body,
            data: data
        )
    }

    // MARK: AppDelegate hooks

    /// APNs reports the device token through `UIApplicationDelegate`
    /// callbacks the host app's generated delegate does not implement. Add
    /// them at runtime (or wrap the originals if a future template adds
    /// them) and forward to the plugin. Installed once, from `load`.
    private static var hooksInstalled = false

    private static func installAppDelegateHooks() {
        guard !hooksInstalled, let delegate = UIApplication.shared.delegate else { return }
        hooksInstalled = true
        let cls: AnyClass = type(of: delegate)

        let registeredSelector = #selector(
            UIApplicationDelegate.application(_:didRegisterForRemoteNotificationsWithDeviceToken:))
        install(on: cls, selector: registeredSelector) { _, token in
            guard let deviceToken = token as? Data else { return }
            DispatchQueue.main.async { shared?.apnsTokenReceived(deviceToken) }
        }

        let failedSelector = #selector(
            UIApplicationDelegate.application(_:didFailToRegisterForRemoteNotificationsWithError:))
        install(on: cls, selector: failedSelector) { _, err in
            let error =
                (err as? Error)
                ?? NSError(
                    domain: "tauri-plugin-push-notifications", code: -1,
                    userInfo: [NSLocalizedDescriptionKey: "unknown registration failure"])
            DispatchQueue.main.async { shared?.apnsRegistrationFailed(error) }
        }
    }

    /// Add a `(UIApplication, AnyObject) -> Void` delegate method to `cls`,
    /// forwarding to `handler`. If the class already implements it (a future
    /// template, or another plugin got there first), chain: the original
    /// runs first, then `handler` — never replaced, never dropped.
    private static func install(
        on cls: AnyClass, selector: Selector,
        handler: @escaping (AnyObject, AnyObject) -> Void
    ) {
        let addedBlock: @convention(block) (AnyObject, AnyObject, AnyObject) -> Void = {
            _, app, arg in
            handler(app, arg)
        }
        let imp = imp_implementationWithBlock(addedBlock)
        if class_addMethod(cls, selector, imp, "v@:@@") { return }
        guard let method = class_getInstanceMethod(cls, selector) else { return }
        typealias Fn = @convention(c) (AnyObject, Selector, AnyObject, AnyObject) -> Void
        let originalFn = unsafeBitCast(method_getImplementation(method), to: Fn.self)
        let wrapper: @convention(block) (AnyObject, AnyObject, AnyObject) -> Void = {
            target, app, arg in
            originalFn(target, selector, app, arg)
            handler(app, arg)
        }
        method_setImplementation(method, imp_implementationWithBlock(wrapper))
    }
}

@available(iOS 15.0, *)
extension UNAuthorizationStatus {
    /// Authorized or provisional — either delivers notifications.
    fileprivate var grantsDelivery: Bool {
        self == .authorized || self == .provisional || self == .ephemeral
    }
}

@_cdecl("init_plugin_push_notifications")
func initPlugin() -> Plugin {
    if #available(iOS 15.0, *) {
        return PushNotificationsPlugin()
    }
    // The consumer app declares minimumSystemVersion 15.0; this branch only
    // pacifies the compiler on older availability paths.
    fatalError("tauri-plugin-push-notifications requires iOS 15+")
}
